src/providers/mod.rs

```diff
@@ -31,6 +35,60 @@ trait Provider {
         body_text: &str,
         body_html: Option<&str>,
     ) -> AppResult<String>;
+
+    fn build_multipart_mime(
```

Comment:
Although I'm 100% in favour of defining this at the generic provider level, I'm not sure about exposing it as part of the `Provider` trait because it's not something we want the calling code to know about. And it doesn't look like `self` is used anyway, so there's no need for it to be a trait method. Could it just be a plain old function and pulled in standalone?

Reply:
Done :)